### DISCUSSION TITLE
Connecting with Multiple Open Haptic Devices

### DISCUSSION THREAD

**Discussion by z6zou** [PlusToolkit/PlusLib#794] [Q&A] [answered]
Hello,
    I'm currently working with two open haptic devices, which I would like to send the stylus and button transformation matrices to Slicer. Following Multimodel tracking examples online, I generated the configuration files as following:

`<PlusConfigurationversion="2.1">`
`  <DataCollection StartupDelaySec="1.0">`
`    <DeviceSet
      Name="PlusServer: Open Haptics device tracking 2 Devices"
      Description="Broadcasting  device stylyus tracking data through OpenIGTLink." />`
`    <Device`
`      Id="TrackerDeviceLeft"`
`      Type="OpenHaptics"`
`      AcquisitionRate="5"`
`      ToolReferenceFrame="BaseLeft"`
`      DeviceName="Left Device"  >`
`      <DataSources>`
`        <DataSource Type="Tool" Id="StylusLeft"  PortName="Stylus" />`
`        <DataSource Type="Tool" Id="ButtonsLeft"  PortName="Buttons" />`
`      </DataSources>`
`      <OutputChannels>`
`        <OutputChannel Id="NeedleStreamLeft">`
`          <DataSource Id="StylusLeft" />`
`          <DataSource Id="ButtonsLeft" />`
`        </OutputChannel>`
`      </OutputChannels>`
`    </Device>`

`    <Device`
`      Id="TrackerDeviceRight"`
`      Type="OpenHaptics"`
`      AcquisitionRate="5"`
`      ToolReferenceFrame="BaseRight"`
`      DeviceName="Right Device"  >`
`      <DataSources>`
`        <DataSource Type="Tool" Id="StylusRight"  PortName="Stylus" />`
`        <DataSource Type="Tool" Id="ButtonsRight"  PortName="Buttons" />`
`      </DataSources>`
`      <OutputChannels>`
`        <OutputChannel Id="NeedleStreamRight">`
`          <DataSource Id="StylusRight" />`
`          <DataSource Id="ButtonsRight" />`
`        </OutputChannel>`
`      </OutputChannels>`
`    </Device>`

`  </DataCollection>`
`  <PlusOpenIGTLinkServer`
`    MaxNumberOfIgtlMessagesToSend="1"`
`    MaxTimeSpentWithProcessingMs="50"`
`    ListeningPort="18944"`
`    OutputChannelId="NeedleStreamLeft" >`
`    <DefaultClientInfo>`
`      <MessageTypes>`
`        <Message Type="TRANSFORM" />`
`      </MessageTypes>`
`      <TransformNames>`
`        <Transform Name="StylusLeftToBaseLeft" />`
`        <Transform Name="ButtonsLeftToBaseLeft" />`
`      </TransformNames>`
`    </DefaultClientInfo>`
`  </PlusOpenIGTLinkServer>`
`  <PlusOpenIGTLinkServer`
`    MaxNumberOfIgtlMessagesToSend="1"`
`    MaxTimeSpentWithProcessingMs="50"`
`    ListeningPort="18889"`
`    OutputChannelId="NeedleStreamRight" >`
`    <DefaultClientInfo>`
`      <MessageTypes>`
`        <Message Type="TRANSFORM" />`
`      </MessageTypes>`
`      <TransformNames>`
`        <Transform Name="StylusRightToBaseRight" />`
`        <Transform Name="ButtonsRightToBaseRight" />`
`      </TransformNames>`
`    </DefaultClientInfo>`
`  </PlusOpenIGTLinkServer>`
`</PlusConfiguration>`

To listen to the two ports, I created two connectors in 3D Slicer. However, it seems like only the transformations in the first device is updating. I tried to swap the order of the devices and confirmed that. I'm wondering if there is anything wrong with my configuration file, and would like to hear any valuable suggestions to achieve this from you.

Best Regards,
Zoe

**Comment by Sunderlandkyl** [comment]
Are there any messages in the Plus log?

**Reply by z6zou** [reply]
Hello, attached is the log file, where I changes the ports to 18944 and 18945. The log seems normal, but the data sending through port 18945 is not updating.
[022621_122424_PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/6051365/022621_122424_PlusLog.txt)

**Comment by adamrankin** [comment]
Everything seems like it should be working, but without the SDK or devices to test with, I am unable to debug. If you are able to make a debug build, we could investigate together.

**Reply by adamrankin** [reply]
@dzenanz Do you know if the openhaptics SDK needs a different setup if multiple devices are used?

**Reply by dzenanz** [reply]
I don't think multiple devices need different setup, but I worked with it few years ago so I am no longer sure. Maybe @sjh26 can confirm?

**Reply by sjh26** [reply]
Multiple devices are supported through PLUS, but it can be difficult to get multiple devices set up in the GeoMagic  / whatever they are calling it setup software.  I often have to do the pairing setup multiple times to get it to work.  Check in the diagnostic software that both are connected correctly.

**Reply by z6zou** [reply]
Thank you very much for your reply. I have rebuilt a debug build here https://github.com/z6zou/PlusBuild-Openhaptics.git. I have tested the connection through the setup software, the connection seems fine to me. The weird part is that when I swap the order of Device in the configuration file, for example, making right device first instead, the first device's data would always send correctly while the second device's data does not update at all.

**Comment by adamrankin** [comment, marked as answer]
@z6zou It seems this might be a configuration issue with the OpenHaptics devices. Could you play around with those settings and see if you can make progress?

**Reply by z6zou** [reply]
Thank you @adamrankin , I will try to dig around to see if I can do anything.
